<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Matteo Sesia (msesia@stanford.edu)" />

<meta name="date" content="2019-05-16" />

<title>Using SNPknock with Genetic Data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Using SNPknock with Genetic Data</h1>
<h4 class="author">Matteo Sesia (<a href="mailto:msesia@stanford.edu">msesia@stanford.edu</a>)</h4>
<h4 class="date">2019-05-16</h4>



<p>This vignette illustrates the basic usage of the <code>SNPknock</code> package in combination with the phasing software <code>fastphase</code> <span class="citation">(Scheet and Stephens 2006)</span> to create knockoffs of unphased genotypes or phased haplotypes <span class="citation">(Sesia, Sabatti, and Candès 2019, <span class="citation">Sesia et al. (2019)</span>)</span>. Since <code>fastphase</code> is not available as an R package, this functionality of <code>SNPknock</code> requires the user to first obtain a copy of <code>fastphase</code>.</p>
<p>To learn more about the application of <code>SNPknock</code> to large genome-wide association studies <span class="citation">(Sesia et al. 2019)</span>, visit: <a href="https://msesia.github.io/knockoffzoom/" class="uri">https://msesia.github.io/knockoffzoom/</a>.</p>
<div id="obtaining-fastphase" class="section level2">
<h2>Obtaining <code>fastphase</code></h2>
<p><code>fastphase</code> is a phasing and imputation tool based on the hidden Markov model described in <span class="citation">(Scheet and Stephens 2006)</span>.</p>
<p>Binary executables for Linux and Mac OS are available from <a href="http://scheet.org/software.html" class="uri">http://scheet.org/software.html</a>.</p>
<p>Before continuing with this tutorial, download the <code>fastphase</code> tarball from the above link and extract the <code>fastphase</code> executable file into a convenient directory (e.g. “~/bin/”).</p>
</div>
<div id="knockoffs-for-unphased-genotypes" class="section level2">
<h2>Knockoffs for unphased genotypes</h2>
<div id="fitting-the-hidden-markov-model-on-genotype-data" class="section level3">
<h3>Fitting the hidden Markov model on genotype data</h3>
<p>A small synthetic dataset of 1454 unphased genotype SNPs from 100 individuals can be found in the package installation directory. We can load it with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(SNPknock)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">X_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genotypes.RData&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">load</span>(X_file)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">table</span>(X)</a></code></pre></div>
<pre><code>## X
##      0      1      2 
## 130548  90112  25540</code></pre>
<p>Below, we show how to fit a hidden Markov model to this data, with the help of <code>fastphase</code>. Since <code>fastphase</code> takes as input genotype sequences in “.inp” format, we must first convert the X matrix by calling <code>writeXtoInp</code>. By default, this function will write onto a temporary file in the R temporary directory.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Convert X into the suitable fastphase input format, write it into a temporary file</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co"># and return the path to that file.</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">Xinp_file =<span class="st"> </span><span class="kw">writeXtoInp</span>(X)</a></code></pre></div>
<p>Assuming that we have already downloaded <code>fastphase</code>, we can call it to fit the hidden Markov model to X.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">fp_path  =<span class="st"> &quot;~/bin/fastphase&quot;</span> <span class="co"># Path to the fastphase executable</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># Call fastphase and return the path to the parameter estimate files</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">fp_outPath =<span class="st"> </span><span class="kw">runFastPhase</span>(fp_path, Xinp_file, <span class="dt">K =</span> <span class="dv">12</span>, <span class="dt">numit =</span> <span class="dv">15</span>)</a></code></pre></div>
<pre><code>## ~/bin/fastphase -Pp -T1 -K12 -g -H-4 -C15 -S1 -o'/tmp/Rtmpd9jSUL/file142d20959abf' /tmp/Rtmpd9jSUL/file142d11a4dc0d.inp</code></pre>
<p>Above, the <code>SNPknock</code> package could not find <code>fastphase</code> because we did not provide the correct path (we cannot include third-party executable files within this package). However, if you install <code>fastphase</code> separately and provide <code>SNPknock</code> with the correct path, this will work.</p>
<p>If the previous step worked for you, you can find the parameter estimates produced by <code>fastphase</code> in the following files:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">r_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_rhat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">alpha_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_alphahat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">theta_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_thetahat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">char_file  =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_origchars&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a></code></pre></div>
<p>Otherwise, for the sake of this tutorial, you can use the example parameter files provided in the package installation directory:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">r_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genotypes_rhat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">alpha_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genotypes_alphahat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">theta_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genotypes_thetahat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">char_file  =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genotypes_origchars&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a></code></pre></div>
<p>Then, we can construct the hidden Markov model with:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">hmm =<span class="st"> </span><span class="kw">loadHMM</span>(r_file, alpha_file, theta_file, char_file)</a></code></pre></div>
</div>
</div>
<div id="generating-knockoff-genotypes" class="section level2">
<h2>Generating knockoff genotypes</h2>
<p>Finally, we can use the hidden Markov model created above to generate knockoffs.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">Xk =<span class="st"> </span><span class="kw">knockoffGenotypes</span>(X, hmm<span class="op">$</span>r, hmm<span class="op">$</span>alpha, hmm<span class="op">$</span>theta)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw">table</span>(Xk)</a></code></pre></div>
<pre><code>## Xk
##      0      1      2 
## 130114  90466  25620</code></pre>
</div>
<div id="knockoffs-for-phased-haplotypes" class="section level2">
<h2>Knockoffs for phased haplotypes</h2>
<div id="fitting-the-hidden-markov-model-on-haplotype-data" class="section level3">
<h3>Fitting the hidden Markov model on haplotype data</h3>
<p>A small synthetic dataset of 1454 phased haplotype SNPs from 100 individuals can be found in the package installation directory. We can load it with:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">library</span>(SNPknock)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">H_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;haplotypes.RData&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="kw">load</span>(H_file)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="kw">table</span>(H)</a></code></pre></div>
<pre><code>## H
##      0      1 
## 351208 141192</code></pre>
<p>Below, we show how to fit a hidden Markov model to this data, with the help of <code>fastphase</code>. Since <code>fastphase</code> takes as input haplotype sequences in “.inp” format, we must first convert the H matrix by calling <code>writeXtoInp</code>. By default, this function will write onto a temporary file in the R temporary directory.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># Convert X into the suitable fastphase input format, write it into a temporary file</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="co"># and return the path to that file.</span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3">Hinp_file =<span class="st"> </span><span class="kw">writeXtoInp</span>(H, <span class="dt">phased =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>Assuming that we have already downloaded <code>fastphase</code>, we can call it to fit the hidden Markov model to X.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">fp_path  =<span class="st"> &quot;~/bin/fastphase&quot;</span> <span class="co"># Path to the fastphase executable</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="co"># Call fastphase and return the path to the parameter estimate files</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3">fp_outPath =<span class="st"> </span><span class="kw">runFastPhase</span>(fp_path, Hinp_file, <span class="dt">K =</span> <span class="dv">12</span>, <span class="dt">numit =</span> <span class="dv">15</span>, <span class="dt">phased =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## ~/bin/fastphase -Pp -T1 -K12 -g -H-4 -C15 -B -S1 -o'/tmp/Rtmpd9jSUL/file142d64b2bf9b' /tmp/Rtmpd9jSUL/file142d41a6e022.inp</code></pre>
<p>Above, the <code>SNPknock</code> package could not find <code>fastphase</code> because we did not provide the correct path (we cannot include third-party executable files within this package). However, if you install <code>fastphase</code> separately and provide <code>SNPknock</code> with the correct path, this will work.</p>
<p>If the previous step worked for you, you can find the parameter estimates produced by <code>fastphase</code> in the following files:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">r_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_rhat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">alpha_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_alphahat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">theta_file =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_thetahat.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb16-4" data-line-number="4">char_file  =<span class="st"> </span><span class="kw">paste</span>(fp_outPath, <span class="st">&quot;_origchars&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</a></code></pre></div>
<p>Otherwise, for the sake of this tutorial, you can use the example parameter files provided in the package installation directory:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">r_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;haplotypes_rhat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">alpha_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;haplotypes_alphahat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">theta_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;haplotypes_thetahat.txt&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">char_file  =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;haplotypes_origchars&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SNPknock&quot;</span>)</a></code></pre></div>
<p>Then, we can construct the hidden Markov model with:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">hmm =<span class="st"> </span><span class="kw">loadHMM</span>(r_file, alpha_file, theta_file, char_file)</a></code></pre></div>
</div>
<div id="generating-knockoff-haplotypes" class="section level3">
<h3>Generating knockoff haplotypes</h3>
<p>Finally, we can use the hidden Markov model created above to generate knockoffs.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">Hk =<span class="st"> </span><span class="kw">knockoffHaplotypes</span>(H, hmm<span class="op">$</span>r, hmm<span class="op">$</span>alpha, hmm<span class="op">$</span>theta)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">table</span>(Hk)</a></code></pre></div>
<pre><code>## Hk
##      0      1 
## 351155 141245</code></pre>
</div>
</div>
<div id="see-also" class="section level2">
<h2>See also</h2>
<p>If you want to see some basic usage of <code>SNPknock</code>, see the <a href="SNPknock.html">introductory vignette</a>.</p>
<p>If you want to learn about <code>SNPknock</code> for large genome-wide association studies <span class="citation">(Sesia et al. 2019)</span>, see <a href="https://msesia.github.io/knockoffzoom/" class="uri">https://msesia.github.io/knockoffzoom/</a>.</p>
<div id="refs" class="references">
<div id="ref-scheet2006">
<p>Scheet, P., and M. Stephens. 2006. “A Fast and Flexible Statistical Model for Large-Scale Population Genotype Data: Applications to Inferring Missing Genotypes and Haplotypic Phase.” <em>Am. J. Hum. Genet.</em> 78:629–44. <a href="https://doi.org/10.1086/502802" class="uri">https://doi.org/10.1086/502802</a>.</p>
</div>
<div id="ref-sesia2019multi">
<p>Sesia, M., E. Katsevich, S. Bates, E. Candès, and C. Sabatti. 2019. “Multi-Resolution Localization of Causal Variants Across the Genome.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/631390" class="uri">https://doi.org/10.1101/631390</a>.</p>
</div>
<div id="ref-sesia2019">
<p>Sesia, M., C. Sabatti, and E. J. Candès. 2019. “Gene Hunting with Hidden Markov Model Knockoffs.” <em>Biometrika</em> 106:1–18. <a href="https://doi.org/10.1093/biomet/asy033" class="uri">https://doi.org/10.1093/biomet/asy033</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
